docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+
+The OpenVPN server running in the shoot cluster is configured as follows:
+
+```
+mode server
+tls-server
+proto tcp4-server
+dev tun0
+
+[...]
+
+server 192.168.123.0 255.255.255.0
+
+push "route 10.243.0.0 255.255.128.0"
+push "route 10.243.128.0 255.255.128.0"
+
+duplicate-cn
+
+key "/srv/secrets/vpn-shoot/tls.key"
+cert "/srv/secrets/vpn-shoot/tls.crt"
+ca "/srv/secrets/vpn-shoot/ca.crt"
+dh "/srv/secrets/dh/dh2048.pem"
+
+tls-auth "/srv/secrets/tlsauth/vpn.tlsauth" 0
+push "route 10.242.0.0 255.255.0.0"
+```
+
+It is a TCP TLS server and configured to automatically assign IP addresses for OpenVPN clients (`server` directive). In addition, it pushes the shoot cluster node-, pod-, and service ranges to the clients running in the seed cluster (`push` directive). 
+
+**Note:** The network mesh spanned by OpenVPN uses the network range `192.168.123.0 - 192.168.123.255`. This network range cannot be used in either shoot-, or seed clusters. If it is used this might cause subtle problem due to network range overlaps. Unfortunately, this appears not to be well documented but this restriction exists since the very beginning. We should clean up this technical debt as part of the excercise.
+
+### Goals
+
+* We intend to supersede the current VPN solution with the solution outlined in this proposal.
+* We intend to remove the code for the konnectivity tunnel once this solution proposal has been validated.
+
+### Non Goals
+
+* The solution is not a low latency, or high throughput solution. As the kube-apiserver to shoot cluster traffic does not demand these properties we do not intend to invest in improvements.
+* We do not intend to provide continuous availability to the shoot-seed VPN connection. We expect the availability to be comparable to the existing solution.
+
+## Proposal
+
+The propsal is depicted in the following diagram:
+
+![alt text](assets/ReversedTunnelVPN.png "Reversed Tunnel VPN")
+
+We have added an OpenVPN server pod (`vpn-seed-server`) to each control plane. This OpenVPN server is configured with the OpenVPN `client-to-client` option that allows clients to talk to each other. The OpenVPN client in the shoot cluster (`vpn-shoot-client`) as well as the Open VPN client sidecars in the seed (`vpn-seed-client`) connect to the OpenVPN server.
+
+The three containers `vpn-seed-server`, `vpn-seed-client`, and `vpn-shoot-client` are new containers and are not related to containers in the github.com/gardener/vpn project. We will create a new project github.com/gardener/vpn2 for these containers. With this solution we intend to supersede the containers from the github.com/gardener/vpn project.
```

Comment:
This can be also simplified - the `kube-apsierver` is already aware that it can use HTTP CONNECT to connect to shoot networks - https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/ 

We can have the following sidecar (envoy-proxy) running next to the `openvpn-server`:

```yaml
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 9443
    listener_filters:
    - name: "envoy.filters.listener.tls_inspector"
      typed_config: {}
    filter_chains:
    - transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          common_tls_context:
            tls_certificates:
            - certificate_chain: { filename: "/etc/tls/server.crt" }
              private_key: { filename: "/etc/tls/server.key" }
            validation_context:
              trusted_ca:
                filename: /etc/tls/ca.crt
      filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains:
              - "*"
              routes:
              - match:
                  connect_matcher: {}
                route:
                  cluster: dynamic_forward_proxy_cluster
                  upgrade_configs:
                  - upgrade_type: CONNECT
                    connect_config: {}
          http_filters:
          - name: envoy.filters.http.dynamic_forward_proxy
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.dynamic_forward_proxy.v3.FilterConfig
              dns_cache_config:
                name: dynamic_forward_proxy_cache_config
                dns_lookup_family: V4_ONLY
          - name: envoy.filters.http.router
          http_protocol_options:
            accept_http_10: true
          upgrade_configs:
            - upgrade_type: CONNECT
  clusters:
  - name: dynamic_forward_proxy_cluster
    connect_timeout: 1s
    lb_policy: CLUSTER_PROVIDED
    cluster_type:
      name: envoy.clusters.dynamic_forward_proxy
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.clusters.dynamic_forward_proxy.v3.ClusterConfig
        dns_cache_config:
          name: dynamic_forward_proxy_cache_config
          dns_lookup_family: V4_ONLY
```

The API Server can be configured with 

```yaml
apiVersion: apiserver.k8s.io/v1alpha1
kind: EgressSelectorConfiguration
egressSelections:
- name: cluster
  connection:
    proxyProtocol: HTTPConnect
    transport:
      tcp:
        url: https://openvpn-server:9443
        tlsConfig:
          caBundle: /srv/kubernetes/ca/ca.crt
          clientCert: /etc/srv/kubernetes/konnectivity-server-client-tls/tls.crt
          clientKey: /etc/srv/kubernetes/konnectivity-server-client-tls/tls.key
- name: {{ if semverCompare "< 1.20" .Values.kubernetesVersion }}master{{ else }}controlplane{{ end }}
    connection:
    proxyProtocol: Direct
- name: etcd
    connection:
    proxyProtocol: Direct
```

The API Server would talk to the `openvpn-server` Pod on port `9443` where the envoy proxy would establish connection to the right host in the Shoot cluster.

This has the following benefits:

- no need for the API sever to have additional init and sidecar container pods (faster start times)
- removes the need for additional linux capabilities on the apiserver pod (better security)

Reply:
This is a good proposal in order to allow address overlaps between seed and shoot clusters. We would however like to validate the current approach as is (e.g. especially making sure that OpenVPN connections are forwarded correctly by envoy) before tackling this.

Reply:
> This is a good proposal in order to allow address overlaps between seed and shoot clusters. We would however like to validate the current approach as is (e.g. especially making sure that OpenVPN connections are forwarded correctly by envoy) before tackling this.

The HTTP PROXY approach is going to simplify the task - when debugging the API server would simply make a HTTP CONNECT without the need for its OpenVPN to have established connection. Also we can configure [access logging](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage) in envoy, to further access logging.

Reply:
done